### PR TITLE
Add support for client_secret.json

### PR DESCRIPTION
**Pull request by pidi3000**
Create Client from a `client_secret.json` file, as provided on the google cloud console.
Makes creating a client easier. 

However this is limited to "web" type client secrets because others can have missing fields, 
for example some "installed" (android, chrome, etc. ) ones don't have a `client_secret` set.

If there is interest I can add support for those as well, it's a pretty easy fix.

Creating a client with 
``` Python
cli = Client(client_secret_path="path/to/client_secret.json")
``` 
will set the `client_id` and `client_secret`.
Also the `DEFAULT_REDIRECT_URI` will be set to the first entry found in the `redirect_uris` list of the json file, if there are any.

I have not looked at it closely but I think the same changes could be added to the Api class.

### CODE REVIEW

**Comment by MerleLiuKun** [comment]
Wow, that's good. thanks for your pr.  

Cloud you format the code and add some tests for this? :)

**Comment by pidi3000** [comment]
I tried formatting using black, similar to how you have it in the github action.

I don't have much experience with pytest or writing tests in general, but I hope this is alright.

File: pyyoutube/client.py [hunk 1]
```diff
@@ -127,6 +137,68 @@ def __init__(
...
+
+        except KeyError as ke:
+            raise PyYouTubeException(
+                ErrorMessage(
+                    status_code=ErrorCode.INVALID_PARAMS,
+                    message=f"client_secret file is missing required key: '{ke.args[0]}'",
+                )
+            )
+
+        except Exception:
```
Comment (pidi3000) [inline pyyoutube/client.py:186]: I don't know if this should catch any  `Exception` and wrap it in a `PyYouTubeException`

**Comment by pidi3000** [comment]
Also at the very end of the `_from_client_secrets_file` function I'm catching any `Exception` and wrapping it in a `PyYouTubeException`.
I'm not sure if that should be left in or removed.

**Comment by MerleLiuKun** [comment]
Hi，I will help and focus on this at after days. :)

**Comment by pidi3000** [comment]
OK cool, thanks.

**Comment by MerleLiuKun** [comment]
Hi, I left some commets. please take a look. :)

**Comment by pidi3000** [comment]
There are no comments showing up for me. 
I might be missing something, but from testing and reading if the comments show as "Pending" you still have to click "Finnish review" in the top right.

**Comment by MerleLiuKun** [comment]
Sory i forget it. :(

**Comment by pidi3000** [comment]
fixed most of it

File: pyyoutube/client.py [hunk 2]
```diff
@@ -144,54 +143,28 @@ def _from_client_secrets_file(self, client_secret_path: str):
...
-                    ErrorMessage(
-                        status_code=ErrorCode.INVALID_PARAMS,
-                        message="Only 'web' client_secret file are supported.",
-                    )
-                )
-
-        except KeyError as ke:
+        # For now only 'web' client_secret files are support,
+        # some 'installed' type files can have missing 'client_secret' key
+        if "web" not in secrets_data:
```
Comment (pidi3000) [inline pyyoutube/client.py:152]: To support 'installed' type secrets we could do : 

``` Python
secrets_type = list(secrets_data.keys())[0] # get fist key name
if secrets_type not in ["web", "installed"]:
    raise ...

```

File: pyyoutube/client.py [hunk 3]
```diff
@@ -144,54 +143,28 @@ def _from_client_secrets_file(self, client_secret_path: str):
...
-        except Exception:
-            # Not sure if catching any other error and
-            # wrapping it in a PyYouTubeException is a good idea
-            raise PyYouTubeException(
-                ErrorMessage(
-                    status_code=ErrorCode.INVALID_PARAMS,
-                    message="Can not load from client_secret file.",
-                )
-            )
+        secrets_data = secrets_data["web"]
```
Comment (pidi3000) [inline pyyoutube/client.py:160]: changes this to:

``` Python
secrets_data = secrets_data[secrets_type]
```

File: pyyoutube/client.py [hunk 4]
```diff
@@ -144,54 +143,28 @@ def _from_client_secrets_file(self, client_secret_path: str):
...
+
+        self.client_id = secrets_data["client_id"]
+        self.client_secret = secrets_data["client_secret"]
```
Comment (pidi3000) [inline pyyoutube/client.py:163]: This could raise a `KeyError`, because some 'installed' type secrets that don't have `client_secret` set

**Review by pidi3000 — COMMENTED** [review]
This is an idea to support 'installed' type secrets.
What do you think about it?

File: pyyoutube/client.py (hunk 4 above)
Comment (MerleLiuKun) [reply pyyoutube/client.py:163]: If not have `client_secret `, the oauth2 flow maybe can not finished. So I think this is necessary parameter.

We can cover this keyError to promot to the user.

File: pyyoutube/client.py (hunk 2 above)
Comment (MerleLiuKun) [reply pyyoutube/client.py:152]: I don't this use `keys()[0]` is a good idea. Maybe we can do this, just check the key,.

```
credentials = None
for secrets_type in ["web", "installed"]:
    if secrets_type in secrets_data:
        credentials = secrets_data[secrets_type]
if not credentials:
    raise PyYouTubeException(xxxx)
# Then to check client id and client secret
```

Comment (pidi3000) [reply pyyoutube/client.py:152]: Good idea

File: pyyoutube/client.py (hunk 4 above)
Comment (pidi3000) [reply pyyoutube/client.py:163]: Should I just let it fail with `KeyError`? 
Or  Raise a  `PyYouTubeException` like this ?
``` Python
if "client_secret" not in secrets_data:
    raise PyYouTubeException(
        ErrorMessage(
            status_code=ErrorCode.MISSING_PARAMS,
            message="file is missing required field 'client_secret'.",
        )
    )

```

Comment (MerleLiuKun) [reply pyyoutube/client.py:163]: I think we can catch the `KeyError`,  And maybe the `client_id` also not exists. :)

Comment (pidi3000) [reply pyyoutube/client.py:163]: Okey, then I'll quickly make the changes

**Comment by pidi3000** [comment]
I'll next add some more tests for unsupported client secret type and missing fields.

File: pyyoutube/client.py [hunk 5]
```diff
@@ -147,24 +147,35 @@ def _from_client_secrets_file(self, client_secret_path: str):
...
             )
 
-        secrets_data = secrets_data["web"]
+        # check for reqiered fields
+        for field in ["client_secret", "client_id"]:
+            if field not in credentials:
+                raise PyYouTubeException(
+                    ErrorMessage(
+                        status_code=ErrorCode.MISSING_PARAMS,
+                        message=f"file is missing required field '{field}'.",
```
Comment (MerleLiuKun) [inline pyyoutube/client.py:169]: fix typo.

message=f"missing required field '{field}'."

**Comment by pidi3000** [comment]
Should I add some explanations to the docs?

**Comment by MerleLiuKun** [comment]
> Should I add some explanations to the docs?

Yes! :)

**Comment by MerleLiuKun** [comment]
Wow, Very thanks for your pr.  it's very useful!

**Comment by pidi3000** [comment]
Happy to contribute

**Pull request merged**